x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/InitialSecurityConfigurationListener.java

```diff
@@ -0,0 +1,191 @@
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.index.engine.VersionConflictEngineException;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.core.security.user.ElasticUser;
+import org.elasticsearch.xpack.core.ssl.SSLService;
+import org.elasticsearch.xpack.security.authc.esnative.NativeUsersStore;
+import org.elasticsearch.xpack.security.enrollment.InternalEnrollmentTokenGenerator;
+import org.elasticsearch.xpack.security.support.SecurityIndexManager;
+
+import java.io.PrintStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import static org.elasticsearch.xpack.security.authc.esnative.ReservedRealm.BOOTSTRAP_ELASTIC_PASSWORD;
+import static org.elasticsearch.xpack.security.tool.CommandUtils.generatePassword;
+
+public class InitialSecurityConfigurationListener implements BiConsumer<SecurityIndexManager.State, SecurityIndexManager.State> {
+
+    private static final String tokenKey = "token";
+    private static final String fingerprintKey = "caCertFingerprint";
+    private static final String passwordKey = "elasticPassword";
+
+    private static final Logger LOGGER = LogManager.getLogger(InitialSecurityConfigurationListener.class);
+    private final NativeUsersStore nativeUsersStore;
+    private final SecurityIndexManager securityIndexManager;
+    private final Environment environment;
+    private final SSLService sslService;
+    private final Client client;
+
+    public InitialSecurityConfigurationListener(
+        NativeUsersStore nativeUsersStore,
+        SecurityIndexManager securityIndexManager,
+        SSLService sslService,
+        Client client,
+        Environment environment
+    ) {
+        this.nativeUsersStore = nativeUsersStore;
+        this.securityIndexManager = securityIndexManager;
+        this.sslService = sslService;
+        this.client = client;
+        this.environment = environment;
+    }
+
+    @Override
+    public void accept(SecurityIndexManager.State previousState, SecurityIndexManager.State currentState) {
+        final PrintStream out = BootstrapInfo.getOriginalStandardOut();
+        // Check if it has been closed, try to write something so that we trigger PrintStream#ensureOpen
+        out.println();
+        if (out.checkError()) {
+            outputOnError(null);
+            return;
+        }
+        if (previousState.equals(SecurityIndexManager.State.UNRECOVERED_STATE)
+            && currentState.equals(SecurityIndexManager.State.UNRECOVERED_STATE) == false
+            && securityIndexManager.indexExists() == false
+            && XPackSettings.ENROLLMENT_ENABLED.get(environment.settings())) {
+            GroupedActionListener<Map<String, String>> groupedActionListener = new GroupedActionListener<>(ActionListener.wrap(results -> {
```

Comment:
Since we're using a GroupedActionListener, we need to have wrap the results of `nativeUsersStore.updateReservedUser` and `enrollmentTokenGenerator.createKibanaEnrollmentToken` into a common type. I selected a Map<String, String> so that I can pass the password from the former but both the token and the ca fingerprint from the latter.